### PR TITLE
Fix/include dependencies fix

### PR DESCRIPTION
**Pull request by MrMeemus**
Updated SMO and SqlScriptPublishModel nuget package which contains fixes:
  - UTF-8 default encoding for the output file.
  - Embedded resources handling when in dot net core (Exposed when scripting dependencies).

Scripting Service Fix:
  - null exception in property selector when matching objects.
  - Duplicates being scripted when scripting dependencies due to missing server name in URN.

### CODE REVIEW

**Comment by pensivebrian** [comment]
Missing the SMO package? I see a delete, but not an add.

---
In reply to: [314908013](https://github.com/Microsoft/sqltoolsservice/pull/410#issuecomment-314908013) [](ancestors = 314908013)

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs [hunk 1]
```diff
@@ -218,16 +217,52 @@ private SqlScriptPublishModel BuildPublishModel()
...
-                publishModel.SelectedObjects.Add(scriptingObject.ToUrn(database));
+                publishModel.SelectedObjects.Add(scriptingObject.ToUrn(server, database));
             }
 
             return publishModel;
         }
 
+        private string GetServerNameFromLiveInstance(string connectionString)
+        {
+            SqlConnection connection = new SqlConnection(connectionString);
```
Comment (pensivebrian) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs]: >SqlConnection connection = new SqlConnection(connectionString); [](start = 12, length = 63)

Using statement

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs [hunk 2]
```diff
@@ -218,16 +217,52 @@ private SqlScriptPublishModel BuildPublishModel()
...
+            SqlCommand cmd = null;
```
Comment (pensivebrian) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs]: Using statement:: using(SqlCommand cmd = connection.CreateCommand()

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs [hunk 3]
```diff
@@ -218,16 +217,52 @@ private SqlScriptPublishModel BuildPublishModel()
...
+                Logger.Write(LogLevel.Error, string.Format("Error during SQL query for server name. Message: {0}", e.Message));
+            }
+            finally
+            {
+                if (cmd != null)
+                {
+                    cmd.Dispose();
+                }
+                connection.Close();
+            }
```
Comment (pensivebrian) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs]: We can remove moving to using statements which cleans things up.

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs [hunk 4]
```diff
@@ -218,16 +217,52 @@ private SqlScriptPublishModel BuildPublishModel()
...
+            try {
+                connection.Open();
+            
+                cmd = connection.CreateCommand();
+                cmd.CommandText = "SELECT @@servername";
+
+                object retValue = cmd.ExecuteScalar();  
+                serverName = retValue as string;
+            }
+            catch(Exception e)
```
Comment (pensivebrian) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs]: >( [](start = 17, length = 1)

nit: space

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs [hunk 5]
```diff
@@ -218,16 +217,52 @@ private SqlScriptPublishModel BuildPublishModel()
...
+            {
+                Logger.Write(LogLevel.Error, string.Format("Error during SQL query for server name. Message: {0}", e.Message));
+            }
```
Comment (pensivebrian) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs]: So we're swallowing the exception. Should we just let it bubble up? Or rethrow?

Comment (MrMeemus) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs]: we should rethrow or else the user won't know, good catch!

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs [hunk 6]
```diff
@@ -75,7 +75,6 @@ public override void Execute()
                 ScriptOutputOptions outputOptions = new ScriptOutputOptions
                 {
                     SaveFileMode = ScriptFileMode.Overwrite,
-                    SaveFileType = ScriptFileType.Unicode,          // UTF-16
```
Comment (pensivebrian) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs]: Can you add a comment above that SMO is hard coded to produce a UTF8 file.

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs [hunk 7]
```diff
@@ -218,16 +217,52 @@ private SqlScriptPublishModel BuildPublishModel()
                     "Scripting object count {0}, objects: {1}",
                     selectedObjects.Count(),
                     string.Join(", ", selectedObjects)));
-
+            
+            // Get the server name explicitly after connecting to the server.
+            // This is required to catch the actual server name when targeting sql server on docker in linux.
```
Comment (pensivebrian) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs]: > in linux [](start = 99, length = 9)

Let's remove..they could be using macOS :)

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs [hunk 8]
```diff
@@ -218,16 +217,52 @@ private SqlScriptPublishModel BuildPublishModel()
...
             return publishModel;
         }
 
+        private string GetServerNameFromLiveInstance(string connectionString)
+        {
+            SqlConnection connection = new SqlConnection(connectionString);
+            SqlCommand cmd = null;
+            string serverName = string.Empty;
+
+            try {
```
Comment (pensivebrian) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs]: > { [](start = 15, length = 2)

nit: brace on newline

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs [hunk 9]
```diff
@@ -218,16 +217,52 @@ private SqlScriptPublishModel BuildPublishModel()
...
+            {
+                if (cmd != null)
+                {
+                    cmd.Dispose();
+                }
+                connection.Close();
+            }
+
+            return serverName;
+        }
```
Comment (pensivebrian) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs]: >} [](start = 8, length = 1)

nit: newline after curly brace

**Comment by pensivebrian** [comment]
:clock1:

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingExtensionMethods.cs [hunk 10]
```diff
@@ -89,7 +89,7 @@ public static List<ScriptingObject> GetDatabaseObjects(this SqlScriptPublishMode
         /// <param name="scriptingObject">The scripting object instance.</param>
         /// <param name="database">The name of the database referenced by the Urn.</param>
         /// <returns>The Urn instance.</returns>
-        public static Urn ToUrn(this ScriptingObject scriptingObject, string database)
+        public static Urn ToUrn(this ScriptingObject scriptingObject, string server, string database)
```
Comment (pensivebrian) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingExtensionMethods.cs]: >server [](start = 77, length = 6)

Need to add a Validate.IsNotNullOrWhitespaceString("server", server);

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingObjectMatcher.cs [hunk 11]
```diff
@@ -226,9 +226,11 @@ private static IEnumerable<ScriptingObject> MatchCriteria(string property, Func<
                 }
                 if (property.EndsWith(Wildcard, StringComparison.OrdinalIgnoreCase))
                 {
-                    matchedObjects = candidates.Where(o => propertySelector(o).StartsWith(
-                        propertySelector(o).Substring(0, propertySelector(o).Length - 1),
-                        StringComparison.OrdinalIgnoreCase));
+                    matchedObjects = candidates.Where(
+                        o => 
+                            propertySelector(o) != null && 
+                            propertySelector(o).StartsWith(propertySelector(o).Substring(0, propertySelector(o).Length - 1),
```
Comment (pensivebrian) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingObjectMatcher.cs]: >propertySelector(o).Substring(0, propertySelector(o).Length - 1), [](start = 59, length = 65)

nit: can you move this to a newline such that it's clear both parameters are being passed to .StartsWith()

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs [hunk 12]
```diff
@@ -218,16 +218,46 @@ private SqlScriptPublishModel BuildPublishModel()
...
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                Logger.Write(LogLevel.Error, string.Format("Error during SQL query for server name. Message: {0}", e.Message));
+                throw e;
+            }           
+
+            return serverName;
```
Comment (pensivebrian) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs]: >                     [](start = 30, length = 20)

nit: extra spaces

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs [hunk 13]
```diff
@@ -218,16 +218,46 @@ private SqlScriptPublishModel BuildPublishModel()
...
+
+                        object retValue = cmd.ExecuteScalar();  
+                        serverName = retValue as string;
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                Logger.Write(LogLevel.Error, string.Format("Error during SQL query for server name. Message: {0}", e.Message));
+                throw e;
```
Comment (pensivebrian) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs]: >e [](start = 22, length = 1)

The 'throw e' statement will reset the callstack. I think you just want 'throw'. However, I think we should remove the try catch - the call stack will be pretty clear in the log so I don't think we need to log a message.

Comment (MrMeemus) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs]: sure will do!

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs (hunk 7 above)
Comment (pensivebrian) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs]: Eeek. The comment was fine...I was just referring to removing the "in linux" part of the comment.

---
In reply to: [127084230](https://github.com/Microsoft/sqltoolsservice/pull/410#discussion_r127084230) [](ancestors = 127084230)

Comment (pensivebrian) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs]: Can you restore the comment?

---
In reply to: [127087410](https://github.com/Microsoft/sqltoolsservice/pull/410#discussion_r127087410) [](ancestors = 127087410,127084230)

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs [hunk 14]
```diff
@@ -218,16 +218,40 @@ private SqlScriptPublishModel BuildPublishModel()
...
             }
 
             return publishModel;
         }
 
+        private string GetServerNameFromLiveInstance(string connectionString)
+        {
+            string serverName = string.Empty;
+            try
+            {
```
Comment (pensivebrian) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs]: You can remove the try {}

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs [hunk 15]
```diff
@@ -218,16 +218,40 @@ private SqlScriptPublishModel BuildPublishModel()
...
-                publishModel.SelectedObjects.Add(scriptingObject.ToUrn(database));
+                publishModel.SelectedObjects.Add(scriptingObject.ToUrn(server, database));
             }
 
             return publishModel;
         }
 
+        private string GetServerNameFromLiveInstance(string connectionString)
+        {
+            string serverName = string.Empty;
```
Comment (pensivebrian) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs]: >string serverName = string.Empty; [](start = 12, length = 33)

nit: you can declare and assign this variable in one statement....removes one line.

Comment (pensivebrian) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs]: provided you move the return into the using statement, which I think makes sense.

---
In reply to: [127088389](https://github.com/Microsoft/sqltoolsservice/pull/410#discussion_r127088389) [](ancestors = 127088389)

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs [hunk 16]
```diff
@@ -218,16 +218,40 @@ private SqlScriptPublishModel BuildPublishModel()
...
+                    using(SqlCommand cmd = connection.CreateCommand())
+                    {
+                        cmd.CommandText = "SELECT @@servername";
+
+                        object retValue = cmd.ExecuteScalar();  
+                        serverName = retValue as string;
+                    }
+                }
+            }   
+            return serverName;
```
Comment (pensivebrian) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs]: >                     [](start = 30, length = 20)

nit: still have trailing spaces

**Comment by pensivebrian** [comment]
LGTM

**Pull request merged**